pkg/boot/fit/fit_test.go

```diff
@@ -62,22 +80,236 @@ func TestLoadConfigMiss(t *testing.T) {
 	}
 
-	i.Kernel, i.InitRAMFS = "kernel@0", "ramdisk@0"
+	for _, tt := range []struct {
+		desc           string
+		kernel         string
+		initram        string
+		keyring        openpgp.KeyRing
+		want           error
+		wantKernelFill string
+		wantInitFill   string
+	}{
+		{
+			desc:           "Successful kernel0/init0 read with key0",
+			keyring:        openpgp.EntityList{keys[0]},
+			kernel:         "kernel@0",
+			initram:        "ramdisk@0",
+			want:           nil,
+			wantKernelFill: kernel0Fill,
+			wantInitFill:   initram0Fill,
+		},
+		{
+			desc:           "Successful unsigned kernel1/init",
+			keyring:        nil,
+			kernel:         "kernel@1",
+			initram:        "ramdisk@0",
+			want:           nil,
+			wantKernelFill: kernel1Fill,
+			wantInitFill:   initram0Fill,
+		},
+		{
+			desc:           "Successful unsigned kernel1",
+			keyring:        nil,
+			kernel:         "kernel@1",
+			initram:        "",
+			want:           nil,
+			wantKernelFill: kernel1Fill,
+			wantInitFill:   "",
+		},
+		{
+			desc:           "bad kernel0 good init0 read with key1",
+			keyring:        openpgp.EntityList{keys[1]},
+			kernel:         "kernel@0",
+			initram:        "ramdisk@0",
+			want:           vfile.ErrUnsigned{},
+			wantKernelFill: "",
+			wantInitFill:   "",
+		},
+		{
+			desc:           "missing kernel",
+			keyring:        nil,
+			kernel:         "",
+			initram:        "ramdisk@0",
+			want:           fmt.Errorf(""),
+			wantKernelFill: "",
+			wantInitFill:   "",
+		},
+	} {
+		t.Run(tt.desc, func(t *testing.T) {
+			i, err := New("testdata/fitimage.itb")
+			if err != nil {
+				t.Fatal(err)
+			}
 
-	defer func(old func(i *boot.LinuxImage, verbose bool) error) { loadImage = old }(loadImage)
+			i.Kernel, i.InitRAMFS, i.KeyRing = tt.kernel, tt.initram, tt.keyring
 
-	loadImage = func(i *boot.LinuxImage, verbose bool) error {
-		t.Log("mock load image")
-		return nil
+			defer func(old func(i *boot.LinuxImage, verbose bool) error) { loadImage = old }(loadImage)
+
+			loadImage = func(i *boot.LinuxImage, verbose bool) error {
+				if i == nil {
+					t.Errorf("Load() of kernel:%s, init:%s, keys: %v - passed nil to loadImage", tt.kernel, tt.initram, tt.keyring)
+					return nil
+				}
+
+				if i.Kernel == nil && tt.wantKernelFill != "" {
+					t.Errorf("loadImage gave nil kernel: want pattern '%s'", tt.wantKernelFill)
+				}
+				if i.Kernel != nil {
+					// Make sure we can only read imageSize by expecting an EOF
+					gotKernel := make([]byte, imageSize+1)
+					gotSize, err := i.Kernel.ReadAt(gotKernel, 0)
+					if !errors.Is(err, io.EOF) {
+						t.Errorf("failed reading kernel passed to loadImage: ReadAt(%d)/expected %d: got %v, want %v", imageSize+1, imageSize, err, io.EOF)
+					}
+					if gotSize != imageSize {
+						t.Errorf("failed reading entire kernel passed to loadImage: bytes got %v, want %v", gotSize, imageSize)
+					}
+					wantKernel := fillPattern(tt.wantKernelFill, imageSize)
```

Comment:
why not check in a for loop that every len(pattern) chunk of gotKernel is equal to fillPattern?

Reply:
The bytes.Equal does effectively check the entire contents. We try to read size+1 just to make sure we hit the end, and compare with image[0:size].
Squashed a comment improvement. PTAL if what it's doing is still abstract

Reply:
I assume this checks the entire contents, but why allocate this much memory when it's not needed?

Reply:
Ahh, thanks for explaining that. Eh, 2n memory in testing seemed fine at the scope of 100 bytes each for cleaner code. But understandable.
Squashed an update to scan each byte instead of the string compare.